editor/plugins/curve_editor_plugin.cpp

```diff
@@ -33,276 +33,354 @@
+				Vector2 new_pos = get_world_pos(mpos).clamp(Vector2(0.0, curve->get_min_value()), Vector2(1.0, curve->get_max_value()));
+				if (snap_enabled || mb->is_ctrl_pressed()) {
+					new_pos = new_pos.snapped(Vector2(1.0 / snap_count, (curve->get_range()) / snap_count));
+				}
+
+				new_pos.x = get_offset_without_collision(selected_index, new_pos.x, mpos.x >= get_view_pos(new_pos).x);
+
+				// Add a temporary point for the user to adjust before adding it permanently.
+				int new_idx = curve->add_point_no_update(new_pos);
+				set_selected_index(new_idx);
+				grabbing = GRAB_ADD;
 			}
 		}
+	}
 
-		if (!mb.is_pressed() && _dragging && mb.get_button_index() == MouseButton::LEFT) {
-			_dragging = false;
-			if (_has_undo_data) {
-				EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-				undo_redo->create_action(_selected_tangent == TANGENT_NONE ? TTR("Modify Curve Point") : TTR("Modify Curve Tangent"));
-				undo_redo->add_do_method(*_curve_ref, "_set_data", _curve_ref->get_data());
-				undo_redo->add_undo_method(*_curve_ref, "_set_data", _undo_data);
-				// Note: this will trigger one more "changed" signal even if nothing changes,
-				// but it's ok since it would have fired every frame during the drag anyways
-				undo_redo->commit_action();
-
-				_has_undo_data = false;
+	if (mb.is_valid() && mb->get_button_index() == MouseButton::LEFT && !mb->is_pressed()) {
+		if (selected_tangent_index != TANGENT_NONE) {
+			// Finish moving a tangent control.
+			if (selected_index == 0) {
+				set_point_right_tangent(selected_index, curve->get_point_right_tangent(selected_index));
+			} else if (selected_index == curve->get_point_count() - 1) {
+				set_point_left_tangent(selected_index, curve->get_point_left_tangent(selected_index));
+			} else {
+				set_point_tangents(selected_index, curve->get_point_left_tangent(selected_index), curve->get_point_right_tangent(selected_index));
 			}
+			grabbing = GRAB_NONE;
+		} else if (grabbing == GRAB_MOVE) {
+			// Finish moving a point.
+			set_point_position(selected_index, curve->get_point_position(selected_index));
+			grabbing = GRAB_NONE;
+		} else if (grabbing == GRAB_ADD) {
+			// Finish inserting a new point. Remove the temporary point and insert a permanent one in its place.
+			Vector2 new_pos = curve->get_point_position(selected_index);
+			curve->remove_point(selected_index);
+			add_point(new_pos);
+			grabbing = GRAB_NONE;
 		}
+		queue_redraw();
 	}
 
-	Ref<InputEventMouseMotion> mm_ref = p_event;
-	if (mm_ref.is_valid()) {
-		const InputEventMouseMotion &mm = **mm_ref;
-
-		Vector2 mpos = mm.get_position();
+	Ref<InputEventMouseMotion> mm = p_event;
+	if (mm.is_valid()) {
+		Vector2 mpos = mm->get_position();
 
-		if (_dragging && _curve_ref.is_valid()) {
-			if (_selected_point != -1) {
-				Curve &curve = **_curve_ref;
+		if (grabbing != GRAB_NONE && curve.is_valid()) {
```

Comment:
I don't know why the check for curve validity is there in the first place, I just copied the old implementation. Maybe there's a reason, but in any case it doesn't seem to be causing bugs so I don't plan to touch it.

Reply:
I do feel like it would make sense to just put it right at the top of this function. I don't think there's currently _any_ input at all that you want to do if there's no curve selected, right? This shouldn't even be open if no curve is selected, right?

Reply:
I don't really know about this... but the else condition handles the null curve condition gracefully, so it can't lead to trouble either way.